src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java

```diff
@@ -306,4 +371,124 @@ public void setBroadleafFileService(BroadleafFileService bfs) {
+    }
+
+    /**
+     * Get MultiTenantSite Class by Reflection 
+     * @return
+     * @throws Throwable 
+     */
+    protected Class<?> getMultiTenantClass() throws ClassNotFoundException, LinkageError  {
+        try {
+            Class<?> c = ClassUtils.forName(MULTITENANT_SITE_CLASSNAME, getClass().getClassLoader());
+            return c;
+        } catch (ClassNotFoundException | LinkageError e) {
+            throw e;
+
+        }
+    }
+
+
+    /**
+     * Gets getParentId method by reflection   
+     * @return
+     * @throws NoSuchMethodException, SecurityException,ClassNotFoundException, LinkageError 
+     */
+    protected Method getMultiTenantGetParentSiteIdMethod() throws NoSuchMethodException, SecurityException,ClassNotFoundException, LinkageError {
+        try {
+            return getMultiTenantClass().getMethod(MULTITENANTSITE_GETPARENTID_METHODNAME);
+        } catch (NoSuchMethodException | SecurityException | ClassNotFoundException | LinkageError  e) {
+            // TODO Auto-generated catch block
+            throw e;
+        }
+    }
+
+
+    /**
+     * Reflection getParentSiteId Invocation 
+     * @param site
+     * @return
+     */
+    protected Long invokeMultiTenantGetParentSiteIdMethod(Object site) {
+        try {
+            Long result =Long.parseLong(
+                    ReflectionUtils.invokeMethod(getMultiTenantGetParentSiteIdMethod(), site).toString());
+            return result;
+        } catch ( SecurityException | IllegalArgumentException | NoSuchMethodException | ClassNotFoundException | LinkageError e) {
+            LOG.error("Problem invoking getParentSiteId method from reflected Multitenant class", e);
+            return 0L;
+
+        }
+        
+    }
+
+
+    /**
+     * Get the parent from a MultiTentantSite.
+     * The MultiTenantSite environment is detected by reflection.
+     * @param resourceName
+     * @return
+     */
+    protected String getMultiTenantSiteSpecificResourceNameParent(String resourceName) {
+        BroadleafRequestContext brc = BroadleafRequestContext.getBroadleafRequestContext();
+        
+        if (brc != null) {
+            //getMultiTenantClass retrieves the class by reflection, the casts.
+            Object site="";
+            try {
+                site = (getMultiTenantClass().cast(brc.getNonPersistentSite()));
+            } catch (ClassNotFoundException | LinkageError e) {
+                //The MultitenantEnvironment should be checked before reaching this point so
+                //this exception should never occur if the MultiTenant environment is ok
+               LOG.error("Problem reflecting Multitenant Class", e);
+            }
+            if (site != null) {
+                //The MultiTenantSite.getParentSiteId method is called by reflection.
+                String siteDirectory = getSiteDirectory((invokeMultiTenantGetParentSiteIdMethod(site)));
```

Comment:
Don't you need to gate this call? If for some reason this invocation failed, you are hardcoding in a result of `0L`, which means you'll look in a directory for `site-0`. Can't you just skip this altogether if there was a problem getting to the parent site id?